source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1160,23 +1225,71 @@ enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
   const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
   const bool client_ocsp_capable = isClientOcspCapable(ssl_client_hello);
-
-  // Fallback on first certificate.
-  const TlsContext* selected_ctx = &tls_contexts_[0];
-  auto ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
-  for (const auto& ctx : tls_contexts_) {
-    if (client_ecdsa_capable != ctx.is_ecdsa_) {
-      continue;
+  absl::string_view sni = absl::NullSafeStringView(
+      SSL_get_servername(ssl_client_hello->ssl, TLSEXT_NAMETYPE_host_name));
+
+  const TlsContext* selected_ctx = nullptr;
+  OcspStapleAction ocsp_staple_action;
+
+  // do SNI matching and pkey type matching if SNI exists
+  if (!sni.empty()) {
+    PkeyTypesMap pkey_types_map;
+    // Match on exact server name, i.e. "www.example.com" for "www.example.com".
+    const auto server_name_exact_match = server_names_map_.find(sni);
+    if (server_name_exact_match != server_names_map_.end()) {
+      pkey_types_map = server_name_exact_match->second;
+    } else {
+      // Match on wildcard domain, i.e. ".example.com" for "www.example.com".
+      // https://datatracker.ietf.org/doc/html/rfc6125#section-6.4
+      size_t pos = sni.find('.', 1);
+      if (pos < sni.size() - 1 && pos != std::string::npos) {
+        absl::string_view wildcard = sni.substr(pos);
+        const auto server_name_wildcard_match =
+            server_names_map_.find(static_cast<std::string>(wildcard));
+        if (server_name_wildcard_match != server_names_map_.end()) {
+          pkey_types_map = server_name_wildcard_match->second;
+        }
+      }
```

Comment:
Don't copy the pkey map by value out of the server names map

```suggestion
    if (server_name_it == server_names_map_.end()) {
      // Match on wildcard domain, i.e. ".example.com" for "www.example.com".
      // https://datatracker.ietf.org/doc/html/rfc6125#section-6.4
      size_t pos = sni.find('.', 1);
      if (pos < sni.size() - 1 && pos != std::string::npos) {
        absl::string_view wildcard = sni.substr(pos);
        server_name_it = server_names_map_.find(static_cast<std::string>(wildcard));
      }
      
      if (server_name_it != server_names_map_.end()) {
        const auto& pkey_types_map = server_name_it->second;
        < Do stuff with pkey_types_map >
```